src/decode.rs

```diff
@@ -5071,6 +5065,7 @@ pub unsafe fn rav1d_submit_frame(c: &mut Rav1dContext) -> Rav1dResult {
                     "Compiled without support for {}-bit decoding",
                     8 + 2 * (*f.seq_hdr).hbd
                 );
+                drop(guard);
                 on_error(f, c, out_delayed);
                 return Err(ENOPROTOOPT);
```

Comment:
I don't think any of these `drop(guard)`s immediately before `on_error(f, c, out_delayed)` and `return ...` are necessary, since `pthread_mutex_unlock` was called as the last statement in `on_error` and so that corresponds to the automatic drop after `return`ing.

Reply:
They're necessary to end the guard lifetime before mutably borrowing `c`. Ideally `on_error` wouldn't need to mutate the context but it looks like it does need to for now. We might be able to remove these later.

Reply:
Ahh.  But is it valid to drop the lock earlier than it was previously?  I assume they were dropped in an order on purpose.

To keep the lock dropping in the same order as before, you can make `on_error` a macro instead, like

```rust
macro_rules! on_error {
        () => {
            // current on_error body
        }
}

on_error!();
```

Reply:
Or more ideally, we could do

```rust
impl Rav1dFrameContext {
    pub unsafe fn on_error(&mut self) {
        self.task_thread.error = AtomicI32::new(1);
        rav1d_cdf_thread_unref(&mut self.in_cdf);
        if (*self.frame_hdr).refresh_context != 0 {
            rav1d_cdf_thread_unref(&mut self.out_cdf);
        }
        for i in 0..7 {
            if !self.refp[i].p.frame_hdr.is_null() {
                rav1d_thread_picture_unref(&mut self.refp[i]);
            }
            rav1d_ref_dec(&mut self.ref_mvs_ref[i]);
        }
        rav1d_picture_unref_internal(&mut self.cur);
        rav1d_thread_picture_unref(&mut self.sr_cur);
        rav1d_ref_dec(&mut self.mvs_ref);
        rav1d_ref_dec(&mut self.seq_hdr_ref);
        rav1d_ref_dec(&mut self.frame_hdr_ref);
        for mut tile in self.tiles.drain(..) {
            rav1d_data_unref_internal(&mut tile.data);
        }
    }
}

...

which keeps most of the `Rav1dFrameContext` logic of dropping things separate and out of the macro.  Also, a bunch of that code has changed so it needs a rebase.
    macro_rules! on_error {
        () => {
            f.on_error();
            if c.n_fc == 1 {
                rav1d_thread_picture_unref(&mut c.out);
            } else {
                rav1d_thread_picture_unref(out_delayed);
            }
            c.cached_error_props = c.in_0.m.clone();
        };
    }
```

Reply:
It's still a bit of a mess as the lock outlives the mutation (although dynamically it doesn't). Fair point on moving the unlock, I'll fix this somehow.

Reply:
The lock only outlives the `&mut` for the whole `c`, but if it's in the same `fn` (the macro way) and can see which fields we access, it knows different ones are accessed so there's no conflict.

Reply:
I've replaced the cached_error_props field with a Mutex. Looking at the other uses, I think we're going to want interior mutability as we tighten mutable borrows of the top-level context (i.e. this field is modified after the context structure is initially populated). I also worked around the borrow of `c.out` by borrowing it as a pointer for now (as we do with `out_delayed`). We're going to have to do something else with that eventually, but we can deal with that when dealing with Rav1dPicture ownership.

I'm not adding a macro here. We shouldn't be complicating the code structure on a slow error path to work around a borrowing issue. It's better to pay the (tiny) bit of runtime work since this is on a slow path (error handling).